### PR TITLE
Refactor and fix solveBvLit for CBQI BV.

### PR DESCRIPTION
**Pull request by aniemetz**
This refactors and simplifies solveBvLit() and fixes the following:
+ generate side conditions for BITVECTOR_NEG, BITVECTOR_NOT, BITVECTOR_PLUS,
   BITVECTOR_XOR over inequalities and disequality
+ fix CONCAT handling (generate side conditions rather then computing an inverse which was incorrect)
+ fix SIGN_EXTEND handling (same as with CONCATs)

### CODE REVIEW

File: src/theory/quantifiers/bv_inverter.cpp [hunk 1]
```diff
@@ -2206,6 +2212,608 @@ static Node getScBvShl(bool pol,
...
+
+  if (idx != 0)
+  {
+    if (idx == 1)
+    {
+      s1 = sv_t[0];
+    }
+    else
+    {
+      for (unsigned i = 0; i < idx; ++i) nbs1 << sv_t[i];
```
Comment (mpreiner) [inline src/theory/quantifiers/bv_inverter.cpp]: Add braces (can still be a one-liner).

File: src/theory/quantifiers/bv_inverter.cpp [hunk 2]
```diff
@@ -2206,6 +2212,608 @@ static Node getScBvShl(bool pol,
...
+
+  if (idx != nchildren-1)
+  {
+    if (idx == nchildren-2)
+    {
+      s2 = sv_t[nchildren-1];
+    }
+    else
+    {
+      for (unsigned i = idx+1; i < nchildren; ++i) nbs2 << sv_t[i];
```
Comment (mpreiner) [inline src/theory/quantifiers/bv_inverter.cpp]: Ditto.

File: src/theory/quantifiers/bv_inverter.cpp [hunk 3]
```diff
@@ -2206,6 +2212,608 @@ static Node getScBvShl(bool pol,
...
+      s2 = nbs2.constructNode();
+    }
+    w2 = bv::utils::getSize(s2);
+    t2 = bv::utils::mkExtract(t, w - w1 - wx - 1, 0);
```
Comment (mpreiner) [inline src/theory/quantifiers/bv_inverter.cpp]: `w2` and `w - w1 - wx - 1` should be the same right? Add assertion and use `w2` instead.

File: src/theory/quantifiers/bv_inverter.cpp [hunk 4]
```diff
@@ -2206,6 +2212,608 @@ static Node getScBvShl(bool pol,
...
+  }
+  else if (litk == BITVECTOR_ULT)
+  {
+    if (s1.isNull())
+    {
+      if (pol)
+      {
+        /* x o s2 < t  (interpret t as tx o t2)
+         * with invertibility condition:
+         * (implies (= tx z) (bvult s2 t2))
```
Comment (mpreiner) [inline src/theory/quantifiers/bv_inverter.cpp]: Use => instead of implies (for consistency).

File: src/theory/quantifiers/bv_inverter.cpp [hunk 5]
```diff
@@ -2206,6 +2212,608 @@ static Node getScBvShl(bool pol,
...
+         * where
+         * z = 0 with getSize(z) = wx  */
+        Node z = bv::utils::mkZero(wx);
+        Node ult = nm->mkNode(BITVECTOR_ULT, s2, t2);
+        scl = nm->mkNode(IMPLIES, tx.eqNode(z), ult);
+      }
+      else
+      {
+        /* x o s2 >= t  (interpret t as tx o t2)
+         * (implies (= tx ones) (bvuge s2 t2))
```
Comment (mpreiner) [inline src/theory/quantifiers/bv_inverter.cpp]: Ditto.

File: src/theory/quantifiers/bv_inverter.cpp [hunk 6]
```diff
@@ -2206,6 +2212,608 @@ static Node getScBvShl(bool pol,
...
+        scl = nm->mkNode(IMPLIES, tx.eqNode(n), uge);
+      }
+    }
+    else if (s2.isNull())
+    {
+      if (pol)
+      {
+        /* s1 o x < t  (interpret t as t1 o tx)
+         * with invertibility condition:
+         * (and (bvule s1 t1) (implies (= s1 t1) (distinct tx z)))
```
Comment (mpreiner) [inline src/theory/quantifiers/bv_inverter.cpp]: Ditto.

File: src/theory/quantifiers/bv_inverter.cpp [hunk 7]
```diff
@@ -2206,6 +2212,608 @@ static Node getScBvShl(bool pol,
...
+    }
+    else
+    {
+      if (pol)
+      {
+        /* s1 o x o s2 < t  (interpret t as t1 o tx o t2)
+         * with invertibility condition:
+         * (and
+         *   (bvule s1 t1)
+         *   (implies (and (= s1 t1) (= tx z)) (bvult s2 t2)))
```
Comment (mpreiner) [inline src/theory/quantifiers/bv_inverter.cpp]: Ditto.

File: src/theory/quantifiers/bv_inverter.cpp [hunk 8]
```diff
@@ -2206,6 +2212,608 @@ static Node getScBvShl(bool pol,
...
+        Node imp = nm->mkNode(IMPLIES, a, nm->mkNode(BITVECTOR_ULT, s2, t2));
+        scl = nm->mkNode(AND, ule, imp);
+      }
+      else
+      {
+        /* s1 o x o s2 >= t  (interpret t as t1 o tx o t2)
+         * with invertibility condition:
+         * (and
+         *   (bvuge s1 t1)
+         *   (implies (and (= s1 t1) (= tx ones)) (bvuge s2 t2)))
```
Comment (mpreiner) [inline src/theory/quantifiers/bv_inverter.cpp]: Ditto.

File: src/theory/quantifiers/bv_inverter.cpp [hunk 9]
```diff
@@ -2206,6 +2212,608 @@ static Node getScBvShl(bool pol,
...
+  }
+  else if (litk == BITVECTOR_UGT)
+  {
+    if (s1.isNull())
+    {
+      if (pol)
+      {
+        /* x o s2 > t  (interpret t as tx o t2)
+         * with invertibility condition:
+         * (implies (= tx ones) (bvugt s2 t2))
```
Comment (mpreiner) [inline src/theory/quantifiers/bv_inverter.cpp]: Ditto.

File: src/theory/quantifiers/bv_inverter.cpp [hunk 10]
```diff
@@ -2206,6 +2212,608 @@ static Node getScBvShl(bool pol,
...
+         * ones = ~0 with getSize(ones) = wx  */
+        Node n = bv::utils::mkOnes(wx);
+        Node ugt = nm->mkNode(BITVECTOR_UGT, s2, t2);
+        scl = nm->mkNode(IMPLIES, tx.eqNode(n), ugt);
+      }
+      else
+      {
+        /* x o s2 <= t  (interpret t as tx o t2)
+         * with invertibility condition:
+         * (implies (= tx z) (bvule s2 t2))
```
Comment (mpreiner) [inline src/theory/quantifiers/bv_inverter.cpp]: Ditto.

File: src/theory/quantifiers/bv_inverter.cpp [hunk 11]
```diff
@@ -2206,6 +2212,608 @@ static Node getScBvShl(bool pol,
...
+    }
+    else
+    {
+      if (pol)
+      {
+        /* s1 o x o s2 > t  (interpret t as t1 o tx o t2)
+         * with invertibility condition:
+         * (and
+         *   (bvuge s1 t1)
+         *   (implies (and (= s1 t1) (= tx ones)) (bvugt s2 t2)))
```
Comment (mpreiner) [inline src/theory/quantifiers/bv_inverter.cpp]: Ditto.

File: src/theory/quantifiers/bv_inverter.cpp [hunk 12]
```diff
@@ -2206,6 +2212,608 @@ static Node getScBvShl(bool pol,
...
+        scl = nm->mkNode(IMPLIES, tx.eqNode(z), ule);
+      }
+    }
+    else if (s2.isNull())
+    {
+      if (pol)
+      {
+        /* s1 o x > t  (interpret t as t1 o tx)
+         * with invertibility condition:
+         * (and (bvuge s1 t1) (implies (= s1 t1) (distinct tx ones)))
```
Comment (mpreiner) [inline src/theory/quantifiers/bv_inverter.cpp]: Ditto.

File: src/theory/quantifiers/bv_inverter.cpp [hunk 13]
```diff
@@ -2206,6 +2212,608 @@ static Node getScBvShl(bool pol,
...
+        Node imp = nm->mkNode(IMPLIES, a, nm->mkNode(BITVECTOR_UGT, s2, t2));
+        scl = nm->mkNode(AND, uge, imp);
+      }
+      else
+      {
+        /* s1 o x o s2 <= t  (interpret t as t1 o tx o t2)
+         * with invertibility condition:
+         * (and
+         *   (bvule s1 t1)
+         *   (implies (and (= s1 t1) (= tx z)) (bvule s2 t2)))
```
Comment (mpreiner) [inline src/theory/quantifiers/bv_inverter.cpp]: Ditto.

File: src/theory/quantifiers/bv_inverter.cpp [hunk 14]
```diff
@@ -2206,6 +2212,608 @@ static Node getScBvShl(bool pol,
...
+  }
+  else if (litk == BITVECTOR_SLT)
+  {
+    if (s1.isNull())
+    {
+      if (pol)
+      {
+        /* x o s2 < t  (interpret t as tx o t2)
+         * with invertibility condition:
+         * (implies (= tx min) (bvult s2 t2))
```
Comment (mpreiner) [inline src/theory/quantifiers/bv_inverter.cpp]: Ditto.

File: src/theory/quantifiers/bv_inverter.cpp [hunk 15]
```diff
@@ -2206,6 +2212,608 @@ static Node getScBvShl(bool pol,
...
+         * where
+         * min is the signed minimum value with getSize(min) = wx  */
+        Node min = bv::utils::mkConst(bv::utils::mkBitVectorMinSigned(wx));
+        Node ult = nm->mkNode(BITVECTOR_ULT, s2, t2);
+        scl = nm->mkNode(IMPLIES, tx.eqNode(min), ult);
+      }
+      else
+      {
+        /* x o s2 >= t  (interpret t as tx o t2)
+         * (implies (= tx max) (bvuge s2 t2))
```
Comment (mpreiner) [inline src/theory/quantifiers/bv_inverter.cpp]: Ditto.

File: src/theory/quantifiers/bv_inverter.cpp [hunk 16]
```diff
@@ -2206,6 +2212,608 @@ static Node getScBvShl(bool pol,
...
+        scl = nm->mkNode(IMPLIES, tx.eqNode(max), uge);
+      }
+    }
+    else if (s2.isNull())
+    {
+      if (pol)
+      {
+        /* s1 o x < t  (interpret t as t1 o tx)
+         * with invertibility condition:
+         * (and (bvsle s1 t1) (implies (= s1 t1) (distinct tx z)))
```
Comment (mpreiner) [inline src/theory/quantifiers/bv_inverter.cpp]: Ditto.

File: src/theory/quantifiers/bv_inverter.cpp [hunk 17]
```diff
@@ -2206,6 +2212,608 @@ static Node getScBvShl(bool pol,
...
+    }
+    else
+    {
+      if (pol)
+      {
+        /* s1 o x o s2 < t  (interpret t as t1 o tx o t2)
+         * with invertibility condition:
+         * (and
+         *   (bvsle s1 t1)
+         *   (implies (and (= s1 t1) (= tx z)) (bvult s2 t2)))
```
Comment (mpreiner) [inline src/theory/quantifiers/bv_inverter.cpp]: Ditto.

File: src/theory/quantifiers/bv_inverter.cpp [hunk 18]
```diff
@@ -2206,6 +2212,608 @@ static Node getScBvShl(bool pol,
...
+        Node imp = nm->mkNode(IMPLIES, a, nm->mkNode(BITVECTOR_ULT, s2, t2));
+        scl = nm->mkNode(AND, sle, imp);
+      }
+      else
+      {
+        /* s1 o x o s2 >= t  (interpret t as t1 o tx o t2)
+         * with invertibility condition:
+         * (and
+         *   (bvsge s1 t1)
+         *   (implies (and (= s1 t1) (= tx ones)) (bvuge s2 t2)))
```
Comment (mpreiner) [inline src/theory/quantifiers/bv_inverter.cpp]: Ditto.

File: src/theory/quantifiers/bv_inverter.cpp [hunk 19]
```diff
@@ -2206,6 +2212,608 @@ static Node getScBvShl(bool pol,
...
+  else
+  {
+    Assert(litk == BITVECTOR_SGT);
+    if (s1.isNull())
+    {
+      if (pol)
+      {
+        /* x o s2 > t  (interpret t as tx o t2)
+         * with invertibility condition:
+         * (implies (= tx max) (bvugt s2 t2))
```
Comment (mpreiner) [inline src/theory/quantifiers/bv_inverter.cpp]: Ditto.

File: src/theory/quantifiers/bv_inverter.cpp [hunk 20]
```diff
@@ -2206,6 +2212,608 @@ static Node getScBvShl(bool pol,
...
+         * max is the signed maximum value with getSize(max) = wx  */
+        Node max = bv::utils::mkConst(bv::utils::mkBitVectorMaxSigned(wx));
+        Node ugt = nm->mkNode(BITVECTOR_UGT, s2, t2);
+        scl = nm->mkNode(IMPLIES, tx.eqNode(max), ugt);
+      }
+      else
+      {
+        /* x o s2 <= t  (interpret t as tx o t2)
+         * with invertibility condition:
+         * (implies (= tx min) (bvule s2 t2))
```
Comment (mpreiner) [inline src/theory/quantifiers/bv_inverter.cpp]: Ditto.

File: src/theory/quantifiers/bv_inverter.cpp [hunk 21]
```diff
@@ -2206,6 +2212,608 @@ static Node getScBvShl(bool pol,
...
+        scl = nm->mkNode(IMPLIES, tx.eqNode(min), ule);
+      }
+    }
+    else if (s2.isNull())
+    {
+      if (pol)
+      {
+        /* s1 o x > t  (interpret t as t1 o tx)
+         * with invertibility condition:
+         * (and (bvsge s1 t1) (implies (= s1 t1) (distinct tx ones)))
```
Comment (mpreiner) [inline src/theory/quantifiers/bv_inverter.cpp]: Ditto.

File: src/theory/quantifiers/bv_inverter.cpp [hunk 22]
```diff
@@ -2206,6 +2212,608 @@ static Node getScBvShl(bool pol,
...
+    }
+    else
+    {
+      if (pol)
+      {
+        /* s1 o x o s2 > t  (interpret t as t1 o tx o t2)
+         * with invertibility condition:
+         * (and
+         *   (bvsge s1 t1)
+         *   (implies (and (= s1 t1) (= tx ones)) (bvugt s2 t2)))
```
Comment (mpreiner) [inline src/theory/quantifiers/bv_inverter.cpp]: Ditto.

File: src/theory/quantifiers/bv_inverter.cpp [hunk 23]
```diff
@@ -2206,6 +2212,608 @@ static Node getScBvShl(bool pol,
...
+        Node imp = nm->mkNode(IMPLIES, a, nm->mkNode(BITVECTOR_UGT, s2, t2));
+        scl = nm->mkNode(AND, sge, imp);
+      }
+      else
+      {
+        /* s1 o x o s2 <= t  (interpret t as t1 o tx o t2)
+         * with invertibility condition:
+         * (and
+         *   (bvsle s1 t1)
+         *   (implies (and (= s1 t1) (= tx z)) (bvule s2 t2)))
```
Comment (mpreiner) [inline src/theory/quantifiers/bv_inverter.cpp]: Ditto.

File: src/options/quantifiers_options [hunk 24]
```diff
@@ -347,6 +347,8 @@ option quantEprMatching --quant-epr-match bool :default true
 # CEGQI for BV
 option cbqiBv --cbqi-bv bool :read-write :default true
   use word-level inversion approach for counterexample-guided quantifier instantiation for bit-vectors
+option cbqiBv2 --cbqi-bv2 bool :read-write :default true :link --cbqi-bv :link-smt cbqi-bv
```
[the remark this answers is not in the document]
Comment (aniemetz) [reply src/options/quantifiers_options]: Just realized that this workaround didn't even work. Found the reason why it originally didn't work. Fixed.

File: src/theory/quantifiers/bv_inverter.cpp (hunk 1 above)
Comment (aniemetz) [reply src/theory/quantifiers/bv_inverter.cpp]: Done

File: src/theory/quantifiers/bv_inverter.cpp (hunk 2 above)
Comment (aniemetz) [reply src/theory/quantifiers/bv_inverter.cpp]: Done

File: src/theory/quantifiers/bv_inverter.cpp (hunk 3 above)
Comment (aniemetz) [reply src/theory/quantifiers/bv_inverter.cpp]: You mean `w2 == w - w1 - wx`, right? Done

File: src/theory/quantifiers/bv_inverter.cpp (hunk 4 above)
Comment (aniemetz) [reply src/theory/quantifiers/bv_inverter.cpp]: Done

File: src/theory/quantifiers/bv_inverter.cpp (hunk 5 above)
Comment (aniemetz) [reply src/theory/quantifiers/bv_inverter.cpp]: Done

File: src/theory/quantifiers/bv_inverter.cpp (hunk 6 above)
Comment (aniemetz) [reply src/theory/quantifiers/bv_inverter.cpp]: Done

File: src/theory/quantifiers/bv_inverter.cpp (hunk 7 above)
Comment (aniemetz) [reply src/theory/quantifiers/bv_inverter.cpp]: Done

File: src/theory/quantifiers/bv_inverter.cpp (hunk 8 above)
Comment (aniemetz) [reply src/theory/quantifiers/bv_inverter.cpp]: Done

File: src/theory/quantifiers/bv_inverter.cpp (hunk 9 above)
Comment (aniemetz) [reply src/theory/quantifiers/bv_inverter.cpp]: Done

File: src/theory/quantifiers/bv_inverter.cpp (hunk 10 above)
Comment (aniemetz) [reply src/theory/quantifiers/bv_inverter.cpp]: Done

File: src/theory/quantifiers/bv_inverter.cpp (hunk 12 above)
Comment (aniemetz) [reply src/theory/quantifiers/bv_inverter.cpp]: Done

File: src/theory/quantifiers/bv_inverter.cpp (hunk 11 above)
Comment (aniemetz) [reply src/theory/quantifiers/bv_inverter.cpp]: Done

File: src/theory/quantifiers/bv_inverter.cpp (hunk 13 above)
Comment (aniemetz) [reply src/theory/quantifiers/bv_inverter.cpp]: Done

File: src/theory/quantifiers/bv_inverter.cpp (hunk 14 above)
Comment (aniemetz) [reply src/theory/quantifiers/bv_inverter.cpp]: Done

File: src/theory/quantifiers/bv_inverter.cpp (hunk 15 above)
Comment (aniemetz) [reply src/theory/quantifiers/bv_inverter.cpp]: Done

File: src/theory/quantifiers/bv_inverter.cpp (hunk 16 above)
Comment (aniemetz) [reply src/theory/quantifiers/bv_inverter.cpp]: Done

File: src/theory/quantifiers/bv_inverter.cpp (hunk 17 above)
Comment (aniemetz) [reply src/theory/quantifiers/bv_inverter.cpp]: Done

File: src/theory/quantifiers/bv_inverter.cpp (hunk 18 above)
Comment (aniemetz) [reply src/theory/quantifiers/bv_inverter.cpp]: Done

File: src/theory/quantifiers/bv_inverter.cpp (hunk 19 above)
Comment (aniemetz) [reply src/theory/quantifiers/bv_inverter.cpp]: Done

File: src/theory/quantifiers/bv_inverter.cpp (hunk 20 above)
Comment (aniemetz) [reply src/theory/quantifiers/bv_inverter.cpp]: Done

File: src/theory/quantifiers/bv_inverter.cpp (hunk 21 above)
Comment (aniemetz) [reply src/theory/quantifiers/bv_inverter.cpp]: Done

File: src/theory/quantifiers/bv_inverter.cpp (hunk 22 above)
Comment (aniemetz) [reply src/theory/quantifiers/bv_inverter.cpp]: Done

File: src/theory/quantifiers/bv_inverter.cpp (hunk 23 above)
Comment (aniemetz) [reply src/theory/quantifiers/bv_inverter.cpp]: Done

**Pull request merged**